File: python/taflow/hilbert_transform_sine_wave.py
"""Persistent Hilbert Transform sine wave (HT_SINE)."""

from typing import Any
import numpy as np
from ._native import HilbertTransformSineWave as _Native
from ._series import as_float64_series


class HilbertTransformSineWave:
    """Stateful HilbertTransformSineWave indicator.
    Parameters are documented by the constructor signature; scalar
    ``append`` returns the current value and ``compute`` returns
    the aligned history with NaN warm-up where applicable.
    """

    def __init__(self, _input: Any | None = None):
        """Initialize this adapter and optionally process the supplied input series.

        Parameters
        ----------
        _input : object
            Input series, scalar parameter, or configuration value for this operation.

        Returns
        -------
        object
            The updated adapter, native value, aligned output array, or execution node.
        """
        self._state = _Native()
        if _input is not None:
            self.extend(_input)

    def append(self, value: float):
        """Append one observation or aligned bar to the native Rust state.

        Parameters
        ----------
        value : object
            Input series, scalar parameter, or configuration value for this operation.

        Returns
        -------
        object
            The updated adapter, native value, aligned output array, or execution node.
        """
        self._state.append(float(value))
        return self

    def extend(self, values: Any):
        """Append aligned input series to the native Rust state.

        Parameters
        ----------
        values : object
            Input series, scalar parameter, or configuration value for this operation.

        Returns
        -------
        object
            The updated adapter, native value, aligned output array, or execution node.
        """
        self._state.extend(as_float64_series(values))
        return self

    def compute(self) -> tuple[np.ndarray, np.ndarray]:
        """Return the aligned output history as a NumPy array.

        Returns
        -------
        object
            The updated adapter, native value, aligned output array, or execution node.
        """
        return self._state.compute()

    @property
    def value(self):
        """Return the latest computed value, or None during warm-up.

        Returns
        -------
        object
            The updated adapter, native value, aligned output array, or execution node.
        """
        return self._state.value

    def reset(self):
        """Execute the reset operation through the native Rust implementation.

        Returns
        -------
        object
            The updated adapter, native value, aligned output array, or execution node.
        """
        self._state.reset()
        return self

File: crates/taflow-core/src/indicators/hilbert_transform_sine_wave.rs
//! Incremental Hilbert Transform sine wave (HT_SINE).

use crate::stream::HilbertTransformDominantCyclePhase;

const DEG2RAD: f64 = std::f64::consts::PI / 180.0;

/// Sine and lead-sine values returned by [`HilbertTransformSineWave`].
#[derive(Clone, Copy, Debug, PartialEq)]
/// Persistent Rust state or aligned output type for `HilbertTransformSineWaveValue`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct HilbertTransformSineWaveValue {
    pub sine: f64,
    pub leadsine: f64,
}

/// Incremental HT_SINE state.
/// Persistent Rust state or aligned output type for `HilbertTransformSineWave`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct HilbertTransformSineWave {
    phase: HilbertTransformDominantCyclePhase,
    value: Option<HilbertTransformSineWaveValue>,
}

impl Default for HilbertTransformSineWave {
    fn default() -> Self {
        Self::new()
    }
}
impl HilbertTransformSineWave {
    /// Computes or updates `new` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn new() -> Self {
        Self {
            phase: HilbertTransformDominantCyclePhase::new(),
            value: None,
        }
    }
    /// Appends one price and returns values after TA-Lib's 63-bar warmup.
    pub fn append(&mut self, input: f64) -> Option<HilbertTransformSineWaveValue> {
        self.value = self
            .phase
            .append(input)
            .map(|phase| HilbertTransformSineWaveValue {
                sine: (phase * DEG2RAD).sin(),
                leadsine: ((phase + 45.0) * DEG2RAD).sin(),
            });
        self.value
    }
    /// Computes or updates `value` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn value(&self) -> Option<HilbertTransformSineWaveValue> {
        self.value
    }
    /// Reset the persistent state and clear the latest value.
    pub fn reset(&mut self) {
        self.phase.reset();
        self.value = None;
    }
}
